File: challenge-1-decentralized-staking/packages/nextjs/database/nftdatabase.ts
import sqlite3 from 'sqlite3';

class NFTDatabase {
  private databaseFile: string;
  private db: sqlite3.Database | null;

  constructor(databaseFile: string) {
    this.databaseFile = databaseFile;
    this.db = null;
  }

  // Open the database connection and create necessary tables
  async open(): Promise<void> {
    return new Promise((resolve, reject) => {
      this.db = new sqlite3.Database(this.databaseFile, async (err) => {
        if (err) {
          console.error("Failed to open database:", err.message);
          reject(err);
        } else {
          console.log("Connected to the SQLite database.");
          try {
            await this.createTables();
            resolve();
          } catch (tableErr) {
            reject(tableErr);
          }
        }
      });
    });
  }

  // Create necessary tables
  private createTables(): Promise<void> {
    return new Promise(async (resolve, reject) => {
      try {
        await this.createTable(`
          CREATE TABLE IF NOT EXISTS marketplace_listings (
            listing_id INTEGER PRIMARY KEY AUTOINCREMENT,
            nft_id TEXT NOT NULL,
            seller_address TEXT NOT NULL,
            price REAL NOT NULL,
            listed_at DATETIME DEFAULT CURRENT_TIMESTAMP
          );
        `);
        await this.createTable(`
          CREATE TABLE IF NOT EXISTS transactions_history (
            transaction_id INTEGER PRIMARY KEY AUTOINCREMENT,
            nft_id TEXT NOT NULL,
            buyer_address TEXT NOT NULL,
            seller_address TEXT NOT NULL,
            price REAL NOT NULL,
            transaction_at DATETIME DEFAULT CURRENT_TIMESTAMP
          );
        `);
        await this.createTable(`
          CREATE TABLE IF NOT EXISTS fresh_mints (
            mint_id INTEGER PRIMARY KEY AUTOINCREMENT,
            nft_id TEXT NOT NULL,
            owner_address TEXT NOT NULL,
            minted_at DATETIME DEFAULT CURRENT_TIMESTAMP
          );
        `);
        await this.createTable(`
          CREATE TABLE IF NOT EXISTS queued_mints (
            queue_id INTEGER PRIMARY KEY AUTOINCREMENT,
            nft_id TEXT NOT NULL,
            owner_address TEXT NOT NULL,
            queued_at DATETIME DEFAULT CURRENT_TIMESTAMP
          );
        `);
        resolve();
      } catch (err) {
        reject(err);
      }
    });
  }

  // Create a single table
  private createTable(sql: string): Promise<void> {
    return new Promise((resolve, reject) => {
      if (!this.db) {
        return reject(new Error("Database not open."));
      }
      this.db.run(sql, (err) => {
        if (err) {
          console.error("Error creating table:", err.message);
          reject(err);
        } else {
          console.log("Table created successfully.");
          resolve();
        }
      });
    });
  }

  // Insert a record
  insert(sql: string, params: any[] = []): Promise<number> {
    return new Promise((resolve, reject) => {
      if (!this.db) {
        return reject(new Error("Database not open."));
      }
      this.db.run(sql, params, function (err) {
        if (err) {
          console.error("Error inserting record:", err.message);
          reject(err);
        } else {
          console.log("Record inserted successfully, ID:", this.lastID);
          resolve(this.lastID);
        }
      });
    });
  }

  // Fetch records
  fetch(sql: string, params: any[] = []): Promise<any[]> {
    return new Promise((resolve, reject) => {
      if (!this.db) {
        return reject(new Error("Database not open."));
      }
      this.db.all(sql, params, (err, rows) => {
        if (err) {
          console.error("Error fetching records:", err.message);
          reject(err);
        } else {
          console.log("Records fetched successfully.");
          resolve(rows);
        }
      });
    });
  }

  // Close the database connection
  close(): Promise<void> {
    return new Promise((resolve, reject) => {
      if (!this.db) {
        return reject(new Error("Database not open."));
      }
      this.db.close((err) => {
        if (err) {
          console.error("Failed to close database:", err.message);
          reject(err);
        } else {
          console.log("Database connection closed.");
          resolve();
        }
      });
    });
  }
}

// Example Usage
(async () => {
  const db = new NFTDatabase('nft_database.db');

  try {
    await db.open();

    // Example: Insert a record into marketplace_listings table
    const insertListingSQL = `
      INSERT INTO marketplace_listings (nft_id, seller_address, price)
      VALUES (?, ?, ?)
    `;
    await db.insert(insertListingSQL, [
      "nft12345",
      "0x1234567890abcdef1234567890abcdef12345678",
      10.5
    ]);

    // Fetch records from marketplace_listings table
    const listings = await db.fetch(`SELECT * FROM marketplace_listings`);
    console.log(listings);

  } catch (err) {
    console.error(err);
  } finally {
    await db.close();
  }
})();

File: api/database/nftdatabase/nftdatabase.go
package nftdatabase

import (
    "context"
    "fmt"
    "time"
    "shellhacks/api/database/accountdatabase"
    "github.com/jackc/pgx/v4/pgxpool"
)

// NFTDatabase represents the database for managing NFT data
type NFTDatabase struct {
    Pool *pgxpool.Pool
}

// NewNFTDatabase initializes a new NFTDatabase instance
func NewNFTDatabase(dbURL string) (*NFTDatabase, error) {
    ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
    defer cancel()

    pool, err := pgxpool.Connect(ctx, dbURL)
    if err != nil {
        return nil, fmt.Errorf("failed to connect to NFT database: %w", err)
    }

    // Initialize tables in the NFT database
    if err := InitializeNFTDatabase(pool); err != nil {
        return nil, fmt.Errorf("failed to initialize NFT database tables: %w", err)
    }

    return &NFTDatabase{Pool: pool}, nil
}

// InitializeNFTDatabase initializes tables for NFTs
func InitializeNFTDatabase(pool *pgxpool.Pool) error {
    ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
    defer cancel()

    createMarketplaceListingsTable := `
        CREATE TABLE IF NOT EXISTS marketplace_listings (
            listing_id SERIAL PRIMARY KEY,
            release_name TEXT NOT NULL,
            seller_address TEXT NOT NULL,
            price REAL NOT NULL,
            image_url TEXT,
            listed_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
        );
    `
    createFreshMintsTable := `
        CREATE TABLE IF NOT EXISTS fresh_mints (
            mint_id SERIAL PRIMARY KEY,
            release_name TEXT NOT NULL,
            owner_address TEXT NOT NULL,
            minted_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
        );
    `
    createQueuedMintsTable := `
        CREATE TABLE IF NOT EXISTS queued_mints (
            queue_id SERIAL PRIMARY KEY,
            release_name TEXT NOT NULL,
            owner_address TEXT NOT NULL,
            queued_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
        );
    `

    // Execute the table creation queries
    queries := []string{
        createMarketplaceListingsTable,
        createFreshMintsTable,
        createQueuedMintsTable,
    }

    for _, query := range queries {
        if _, err := pool.Exec(ctx, query); err != nil {
            return fmt.Errorf("failed to execute query (%s): %w", query, err)
        }
    }

    return nil
}

// InsertListing inserts a new listing into the marketplace_listings table
func (db *NFTDatabase) InsertListing(nftID, sellerAddress string, price float64, imageURL string) error {
    ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
    defer cancel()

    query := `
        INSERT INTO marketplace_listings (nft_id, seller_address, price, image_url)
        VALUES ($1, $2, $3, $4)
    `
    _, err := db.Pool.Exec(ctx, query, nftID, sellerAddress, price, imageURL)
    if err != nil {
        return fmt.Errorf("failed to insert listing: %w", err)
    }

    return nil
}

// GetAllListings fetches all marketplace listings from the database
func (db *NFTDatabase) GetAllListings() ([]map[string]interface{}, error) {
    ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
    defer cancel()

    rows, err := db.Pool.Query(ctx, "SELECT * FROM marketplace_listings")
    if err != nil {
        return nil, fmt.Errorf("failed to fetch listings: %w", err)
    }
    defer rows.Close()

    listings := []map[string]interface{}{}

    for rows.Next() {
        var listingID int
        var nftID, sellerAddress, imageURL string
        var price float64
        var listedAt time.Time

        if err := rows.Scan(&listingID, &nftID, &sellerAddress, &price, &imageURL, &listedAt); err != nil {
            return nil, fmt.Errorf("failed to scan row: %w", err)
        }

        listings = append(listings, map[string]interface{}{
            "listing_id":     listingID,
            "release_name":   nftID,
            "seller_address": sellerAddress,
            "price":          price,
            "image_url":      imageURL,
            "listed_at":      listedAt,
        })
    }

    return listings, nil
}

func (db *NFTDatabase) QueueMint(request accountdatabase.ReleaseRequest) error {
    ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
    defer cancel()

    query := `
        INSERT INTO queued_mints (release_name, owner_address)
        VALUES ($1, $2)
    `
    // Assuming nft_id is generated from the release title
    releaseName := request.ReleaseTitle
    _, err := db.Pool.Exec(ctx, query, releaseName, request.Username)
    if err != nil {
        return fmt.Errorf("failed to queue mint: %w", err)
    }

    return nil
}
